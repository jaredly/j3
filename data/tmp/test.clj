100

(fn [x] 101)

(deftype (, a b)
    (, a b))

(defn xx [a b (, c d)]
    (match a
        (, a b) 2
        _       3))

(xx (, 1 2) 3 (, 4 5))

(deftype faces
    (red)
        (black))

red

(def x 101)

((fn [x] 10) 102)

(let [x 11] x)

(def unused 10)

(let [
    x 1
    y 2]
    x)

int-to-string

x

,,

(deftype (list a)
    (nil)
        (cons a (list a)))

(defn map [f lst]
    (match lst
        []           []
        [one ..rest] [(f one) ..(map f rest)]))

(map (+ 2) [1 2 3])