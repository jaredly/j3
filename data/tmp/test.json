{"map":{"0":{"type":"identifier","text":"100","loc":0},"2":{"type":"list","values":[4,5,7],"loc":2},"4":{"type":"identifier","text":"fn","loc":4},"5":{"type":"array","values":[6],"loc":5},"6":{"type":"identifier","text":"x","loc":6},"7":{"type":"identifier","text":"101","loc":7},"16":{"type":"identifier","loc":16,"text":"x"},"18":{"type":"list","values":[20,21,25],"loc":18},"20":{"type":"identifier","text":"deftype","loc":20},"21":{"type":"list","values":[22,23,24],"loc":21},"22":{"type":"identifier","text":",","loc":22},"23":{"type":"identifier","text":"a","loc":23},"24":{"type":"identifier","text":"b","loc":24},"25":{"type":"list","values":[26,28,29],"loc":25},"26":{"type":"identifier","text":",","loc":26},"28":{"type":"identifier","text":"a","loc":28},"29":{"type":"identifier","text":"b","loc":29},"32":{"type":"list","values":[35,36,37],"loc":32},"35":{"type":"identifier","text":"def","loc":35},"36":{"type":"identifier","text":"x","loc":36},"37":{"type":"identifier","text":"101","loc":37},"38":{"type":"list","values":[40,41,44],"loc":38},"40":{"type":"identifier","text":"fn","loc":40},"41":{"type":"array","values":[42],"loc":41},"42":{"type":"identifier","text":"x","loc":42},"44":{"type":"identifier","text":"10","loc":44},"45":{"type":"list","values":[38,46],"loc":45},"46":{"type":"identifier","text":"102","loc":46},"47":{"type":"list","values":[49,50,53],"loc":47},"49":{"type":"identifier","text":"let","loc":49},"50":{"type":"array","values":[51,52],"loc":50},"51":{"type":"identifier","text":"x","loc":51},"52":{"type":"identifier","text":"11","loc":52},"53":{"type":"identifier","text":"x","loc":53},"54":{"type":"identifier","text":"int-to-string","loc":54},"56":{"type":"identifier","text":",,","loc":56},"58":{"type":"list","values":[60,63,66,69],"loc":58},"60":{"type":"identifier","text":"deftype","loc":60},"63":{"type":"list","values":[64,65],"loc":63},"64":{"type":"identifier","text":"list","loc":64},"65":{"type":"identifier","text":"a","loc":65},"66":{"type":"list","values":[68],"loc":66},"68":{"type":"identifier","text":"nil","loc":68},"69":{"type":"list","values":[70,71,72],"loc":69},"70":{"type":"identifier","text":"cons","loc":70},"71":{"type":"identifier","text":"a","loc":71},"72":{"type":"list","values":[73,74],"loc":72},"73":{"type":"identifier","text":"list","loc":73},"74":{"type":"identifier","text":"a","loc":74},"75":{"type":"list","values":[77,78,79,82],"loc":75},"77":{"type":"identifier","text":"defn","loc":77},"78":{"type":"identifier","text":"map","loc":78},"79":{"type":"array","values":[80,81],"loc":79},"80":{"type":"identifier","text":"f","loc":80},"81":{"type":"identifier","text":"lst","loc":81},"82":{"type":"list","values":[83,84,85,86,87,93],"loc":82},"83":{"type":"identifier","text":"match","loc":83},"84":{"type":"identifier","text":"lst","loc":84},"85":{"type":"array","values":[],"loc":85},"86":{"type":"array","values":[],"loc":86},"87":{"type":"array","values":[88,92],"loc":87},"88":{"type":"identifier","text":"one","loc":88},"89":{"type":"identifier","text":"rest","loc":89},"92":{"type":"spread","contents":89,"loc":92},"93":{"type":"array","values":[94,100],"loc":93},"94":{"type":"list","values":[95,96],"loc":94},"95":{"type":"identifier","text":"f","loc":95},"96":{"type":"identifier","text":"one","loc":96},"97":{"type":"list","values":[101,102,103],"loc":97},"100":{"type":"spread","contents":97,"loc":100},"101":{"type":"identifier","text":"map","loc":101},"102":{"type":"identifier","text":"f","loc":102},"103":{"type":"identifier","text":"rest","loc":103},"104":{"type":"list","values":[106,107,110],"loc":104},"106":{"type":"identifier","text":"map","loc":106},"107":{"type":"list","values":[108,109],"loc":107},"108":{"type":"identifier","text":"+","loc":108},"109":{"type":"identifier","text":"2","loc":109},"110":{"type":"array","values":[111,112,114],"loc":110},"111":{"type":"identifier","text":"1","loc":111},"112":{"type":"identifier","text":"2","loc":112},"114":{"type":"identifier","text":"3","loc":114},"115":{"type":"list","values":[117,118,119],"loc":115},"117":{"type":"identifier","text":"def","loc":117},"118":{"type":"identifier","text":"unused","loc":118},"119":{"type":"identifier","text":"10","loc":119},"120":{"type":"list","values":[122,123,128],"loc":120},"122":{"type":"identifier","text":"let","loc":122},"123":{"type":"array","values":[124,125,126,127],"loc":123},"124":{"type":"identifier","text":"x","loc":124},"125":{"type":"identifier","text":"1","loc":125},"126":{"type":"identifier","text":"y","loc":126},"127":{"type":"identifier","text":"2","loc":127},"128":{"type":"identifier","text":"x","loc":128},"129":{"type":"list","values":[131,132,133,135],"loc":129},"131":{"type":"identifier","text":"deftype","loc":131},"132":{"type":"identifier","text":"faces","loc":132},"133":{"type":"list","values":[134],"loc":133},"134":{"type":"identifier","text":"red","loc":134},"135":{"type":"list","values":[136],"loc":135},"136":{"type":"identifier","text":"black","loc":136},"137":{"type":"identifier","text":"red","loc":137},"139":{"type":"list","values":[141,142,143,150],"loc":139},"141":{"type":"identifier","text":"defn","loc":141},"142":{"type":"identifier","text":"xx","loc":142},"143":{"type":"array","values":[144,145,146],"loc":143},"144":{"type":"identifier","text":"a","loc":144},"145":{"type":"identifier","text":"b","loc":145},"146":{"type":"list","values":[147,148,149],"loc":146},"147":{"type":"identifier","text":",","loc":147},"148":{"type":"identifier","text":"c","loc":148},"149":{"type":"identifier","text":"d","loc":149},"150":{"type":"list","values":[154,155,156,160,161,162],"loc":150},"154":{"type":"identifier","text":"match","loc":154},"155":{"type":"identifier","text":"a","loc":155},"156":{"type":"list","values":[157,158,159],"loc":156},"157":{"type":"identifier","text":",","loc":157},"158":{"type":"identifier","text":"a","loc":158},"159":{"type":"identifier","text":"b","loc":159},"160":{"type":"identifier","text":"2","loc":160},"161":{"type":"identifier","text":"_","loc":161},"162":{"type":"identifier","text":"3","loc":162},"163":{"type":"list","values":[165,166,170,171],"loc":163},"165":{"type":"identifier","text":"xx","loc":165},"166":{"type":"list","values":[167,168,169],"loc":166},"167":{"type":"identifier","text":",","loc":167},"168":{"type":"identifier","text":"1","loc":168},"169":{"type":"identifier","text":"2","loc":169},"170":{"type":"identifier","text":"3","loc":170},"171":{"type":"list","values":[172,173,174],"loc":171},"172":{"type":"identifier","text":",","loc":172},"173":{"type":"identifier","text":"4","loc":173},"174":{"type":"identifier","text":"5","loc":174},"-1":{"type":"list","values":[0],"loc":-1}},"root":-1,"meta":{"158":{"trace":{}},"159":{"trace":{}},"163":{"traceTop":{}}},"history":[],"clipboard":[],"hover":[],"regs":{},"at":[{"start":[{"type":"card","idx":-1,"card":0},{"type":"ns","child":140,"idx":0},{"type":"ns-top","idx":140},{"idx":139,"type":"child","at":3},{"idx":150,"type":"child","at":2},{"idx":156,"type":"child","at":1},{"idx":158,"type":"subtext","at":0}]}],"cards":[{"path":[],"top":0}],"nsMap":{"0":{"id":0,"top":-1,"children":[1,3,19,140,164,130,138,33,39,48,116,121,55,17,57,59,76,105],"type":"normal"},"1":{"id":1,"top":0,"children":[],"type":"normal"},"3":{"type":"normal","top":2,"children":[],"id":3},"17":{"type":"normal","top":16,"children":[],"id":17},"19":{"type":"normal","top":18,"children":[],"id":19},"33":{"type":"normal","top":32,"children":[],"id":33},"39":{"type":"normal","top":45,"children":[],"id":39},"48":{"type":"normal","top":47,"children":[],"id":48},"55":{"type":"normal","top":54,"children":[],"id":55},"57":{"type":"normal","top":56,"children":[],"id":57},"59":{"type":"normal","top":58,"children":[],"id":59},"76":{"type":"normal","top":75,"children":[],"id":76},"105":{"type":"normal","top":104,"children":[],"id":105},"116":{"type":"normal","top":115,"children":[],"id":116},"121":{"type":"normal","top":120,"children":[],"id":121},"130":{"type":"normal","top":129,"children":[],"id":130},"138":{"type":"normal","top":137,"children":[],"id":138},"140":{"type":"normal","top":139,"children":[],"id":140,"collapsed":"pinned"},"164":{"type":"normal","top":163,"children":[],"id":164}},"highlight":[158],"evaluator":["parse-1-args.js","algw-list.js","jcst.js"]}