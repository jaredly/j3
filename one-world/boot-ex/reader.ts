import { IDRef, Loc, RecNode } from '../shared/nodes';

type RNode = RecNode;

export const reader = (
    text: string,
    top: string = 'test-top',
    globals: Record<string, IDRef> = {},
): RNode | void => {
    let i = 0;
    const pairs = { '[': ']', '{': '}', '(': ')' };
    const skipWhite = () => {
        while (i < text.length && text[i].match(/\s/)) i++;
        return i >= text.length;
    };
    const l = (num: number): Loc => [[top, num]];

    const readString = (): Extract<RNode, { type: 'string' }> | void => {
        const loc = i;
        i++;
        let first = '';
        let current = '';
        let templates: { suffix: string; expr: RNode; loc: number }[] = [];

        const next = () => {
            if (!templates.length) {
                first = current;
            } else {
                templates[templates.length - 1].suffix = current;
            }
            current = '';
        };

        for (; i < text.length; i++) {
            if (text[i] === '"') {
                i++;
                break;
            }
            if (text[i] === '\\') {
                current += text[i];
                i++;
            }
            if (text[i] === '$' && text[i + 1] === '{') {
                next();
                i += 2;
                const expr = read();
                if (!expr) return;
                templates.push({ expr, suffix: '', loc: i });
                if (text[i] !== '}')
                    throw new Error('unmatched } in template string');
                continue;
            }
            current += text[i];
            // console.log('a', current, text[i])
        }
        next();
        return {
            first,
            templates,
            type: 'string',
            loc: l(loc),
            tag: { type: 'id', text: '', loc: l(loc) },
        };
    };

    const stops = `|"]}) \t\n`;
    const read = (): RNode | void => {
        if (skipWhite()) return;
        if (text[i] === '(' && text[i + 1] === '|') {
            const start = i;
            i += 2;
            const rows: RNode[][] = [[]];
            while (true) {
                if (text[i] === '\n') {
                    rows.push([]);
                    i++;
                }
                if (skipWhite()) return;
                if (text[i] === '|' && text[i + 1] === ')') {
                    break;
                }
                if (text[i] === '|') {
                    i++;
                }
                const next = read();
                if (!next) break;
                rows[rows.length - 1].push(next);
            }
            return {
                type: 'table',
                kind: '(',
                rows,
                loc: l(start),
            };
        }

        switch (text[i]) {
            case '[':
            case '{':
            case '(':
                const start = i;
                const last = pairs[text[i] as '['];
                i += 1;
                const items: RNode[] = [];
                while (true) {
                    const next = read();
                    if (!next) break;
                    items.push(next);
                }
                skipWhite();
                if (i >= text.length || text[i] !== last) {
                    throw new Error(`Expected ${last}: ${text[i]}`);
                }
                i++;
                return {
                    type:
                        last === ']'
                            ? 'array'
                            : last === ')'
                            ? 'list'
                            : 'record',
                    items,
                    loc: l(start),
                };
            case '"':
                return readString();
            case '.':
                if (text[i + 1] === '.') {
                    const loc = l(i);
                    i += 2;
                    const inner = read();
                    if (!inner) return;
                    return { type: 'spread', contents: inner, loc };
                }
            default: {
                let start = i;
                for (; i < text.length && !stops.includes(text[i]); i++) {}
                if (start === i) return;
                let idText = text.slice(start, i);
                if (idText.includes('#')) {
                    const [left, right] = idText.split('#');
                    idText = left;
                    start = +right;
                }
                let ref: IDRef;
                if (idText.includes('!')) {
                    const [left, right] = idText.split('!');
                    idText = left;
                    const key = right.length ? right : left;
                    if (left === '') {
                        ref = { type: 'placeholder', text: right };
                    } else {
                        ref = globals[key];
                        if (!globals[key]) {
                            throw new Error(`global missing: ${key}`);
                        }
                    }
                }
                const id: RNode = {
                    type: 'id',
                    text: idText,
                    loc: l(start),
                    ref,
                };
                if (text[i] === '"') {
                    const string = readString();
                    if (!string) return string;
                    return { ...string, tag: id };
                }
                return id;
            }
        }
    };
    return read();
};
