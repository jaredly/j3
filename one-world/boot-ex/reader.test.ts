// testing the reader I guess

import { reader } from './reader';
import { test, expect } from 'bun:test';

test('ok', () => {
    expect(reader('()')).toMatchSnapshot();
    expect(reader('(1 2 3)')).toMatchSnapshot();
    expect(reader('(hi "folks")')).toMatchSnapshot();
    expect(reader('what"things"')).toMatchSnapshot();
    expect(reader('(hi "folks" and "such${is}life")')).toMatchSnapshot();

    expect(reader('["a\nb" "a\\nb" "a\\\\nb"]')).toMatchSnapshot();
});
