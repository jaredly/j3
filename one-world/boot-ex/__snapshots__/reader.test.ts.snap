// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ok 1`] = `
{
  "items": [],
  "loc": 0,
  "type": "list",
}
`;

exports[`ok 2`] = `
{
  "items": [
    {
      "loc": 1,
      "text": "1",
      "type": "id",
    },
    {
      "loc": 3,
      "text": "2",
      "type": "id",
    },
    {
      "loc": 5,
      "text": "3",
      "type": "id",
    },
  ],
  "loc": 0,
  "type": "list",
}
`;

exports[`ok 3`] = `
{
  "items": [
    {
      "loc": 1,
      "text": "hi",
      "type": "id",
    },
    {
      "first": "folks",
      "loc": 4,
      "tag": {
        "loc": -1,
        "text": "",
        "type": "id",
      },
      "templates": [],
      "type": "string",
    },
  ],
  "loc": 0,
  "type": "list",
}
`;

exports[`ok 4`] = `
{
  "first": "things",
  "loc": 4,
  "tag": {
    "loc": 0,
    "text": "what",
    "type": "id",
  },
  "templates": [],
  "type": "string",
}
`;

exports[`ok 5`] = `
{
  "items": [
    {
      "loc": 1,
      "text": "hi",
      "type": "id",
    },
    {
      "first": "folks",
      "loc": 4,
      "tag": {
        "loc": -1,
        "text": "",
        "type": "id",
      },
      "templates": [],
      "type": "string",
    },
    {
      "loc": 12,
      "text": "and",
      "type": "id",
    },
    {
      "first": "such",
      "loc": 16,
      "tag": {
        "loc": -1,
        "text": "",
        "type": "id",
      },
      "templates": [
        {
          "expr": {
            "loc": 23,
            "text": "is",
            "type": "id",
          },
          "loc": 25,
          "suffix": "life",
        },
      ],
      "type": "string",
    },
  ],
  "loc": 0,
  "type": "list",
}
`;

exports[`ok 6`] = `
{
  "items": [
    {
      "first": 
"a
b"
,
      "loc": 1,
      "tag": {
        "loc": -1,
        "text": "",
        "type": "id",
      },
      "templates": [],
      "type": "string",
    },
    {
      "first": "a\nb",
      "loc": 7,
      "tag": {
        "loc": -1,
        "text": "",
        "type": "id",
      },
      "templates": [],
      "type": "string",
    },
    {
      "first": "a\\nb",
      "loc": 14,
      "tag": {
        "loc": -1,
        "text": "",
        "type": "id",
      },
      "templates": [],
      "type": "string",
    },
  ],
  "loc": 0,
  "type": "array",
}
`;
